File: CS/Scaffolding.ManyToMany/Common/ViewModel/AddRemoveDetailEntitiesViewModel.cs
using DevExpress.Mvvm;
using DevExpress.Mvvm.POCO;
using Scaffolding.ManyToMany.Common.DataModel;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows;

namespace Scaffolding.ManyToMany.Common.ViewModel {
    public class AddRemoveDetailEntitiesViewModel<TEntity, TPrimaryKey, TDetailEntity, TDetailPrimaryKey, TUnitOfWork> : SingleObjectViewModelBase<TEntity, TPrimaryKey, TUnitOfWork>
        where TEntity : class
        where TDetailEntity : class
        where TUnitOfWork : IUnitOfWork {

        public static AddRemoveDetailEntitiesViewModel<TEntity, TPrimaryKey, TDetailEntity, TDetailPrimaryKey, TUnitOfWork> Create(IUnitOfWorkFactory<TUnitOfWork> unitOfWorkFactory, Func<TUnitOfWork, IRepository<TEntity, TPrimaryKey>> getRepositoryFunc, Func<TUnitOfWork, IRepository<TDetailEntity, TDetailPrimaryKey>> getDetailsRepositoryFunc, Func<TEntity, ICollection<TDetailEntity>> getDetailsFunc, TPrimaryKey id) {
            return ViewModelSource.Create(() => new AddRemoveDetailEntitiesViewModel<TEntity, TPrimaryKey, TDetailEntity, TDetailPrimaryKey, TUnitOfWork>(unitOfWorkFactory, getRepositoryFunc, getDetailsRepositoryFunc, getDetailsFunc, id));
        }

        readonly TPrimaryKey id;
        readonly Func<TUnitOfWork, IRepository<TDetailEntity, TDetailPrimaryKey>> getDetailsRepositoryFunc;
        readonly Func<TEntity, ICollection<TDetailEntity>> getDetailsFunc;

        protected IDialogService DialogService { get { return this.GetRequiredService<IDialogService>(); } }

        IRepository<TDetailEntity, TDetailPrimaryKey> DetailsRepository { get { return getDetailsRepositoryFunc(UnitOfWork); } }

        public ICollection<TDetailEntity> DetailEntities { get { return Entity != null ? getDetailsFunc(Entity) : Enumerable.Empty<TDetailEntity>().ToArray(); } }
        public ObservableCollection<TDetailEntity> SelectedEntities { get; private set; }

        protected AddRemoveDetailEntitiesViewModel(IUnitOfWorkFactory<TUnitOfWork> unitOfWorkFactory, Func<TUnitOfWork, IRepository<TEntity, TPrimaryKey>> getRepositoryFunc, Func<TUnitOfWork, IRepository<TDetailEntity, TDetailPrimaryKey>> getDetailsRepositoryFunc, Func<TEntity, ICollection<TDetailEntity>> getDetailsFunc, TPrimaryKey id)
            : base(unitOfWorkFactory, getRepositoryFunc, null) {
            this.id = id;
            this.getDetailsRepositoryFunc = getDetailsRepositoryFunc;
            this.getDetailsFunc = getDetailsFunc;
            SelectedEntities = new ObservableCollection<TDetailEntity>();
            ReloadDetails();
        }

        protected override void OnInitializeInRuntime() {
            base.OnInitializeInRuntime();
            Messenger.Default.Register<EntityMessage<TEntity, TPrimaryKey>>(this, OnMessage);
        }

        public void AddDetailEntities() {
            var availalbleEntities = DetailsRepository.ToList().Except(DetailEntities).ToArray();
            var selectEntitiesViewModel = new SelectDetailEntitiesViewModel<TDetailEntity>(availalbleEntities);
            if(DialogService.ShowDialog(MessageButton.OKCancel, "Select objects to add", selectEntitiesViewModel) == MessageResult.OK && selectEntitiesViewModel.SelectedEntities.Any()) {
                foreach(var entity in selectEntitiesViewModel.SelectedEntities) {
                    DetailEntities.Add(entity);
                }
                SaveChangesAndNotify(selectEntitiesViewModel.SelectedEntities);
            }
        }

        public bool CanAddDetailEntities() {
            return Entity != null;
        }

        public void RemoveDetailEntities() {
            if(!SelectedEntities.Any())
                return;
            foreach(var entity in SelectedEntities) {
                DetailEntities.Remove(entity);
            }
            SaveChangesAndNotify(SelectedEntities);
            SelectedEntities.Clear();
        }

        public bool CanRemoveDetailEntities() {
            return Entity != null;
        }

        void SaveChangesAndNotify(IEnumerable<TDetailEntity> detailEntities) {
            try {
                UnitOfWork.SaveChanges();
                foreach(var entity in detailEntities) {
                    Messenger.Default.Send(new EntityMessage<TDetailEntity, TDetailPrimaryKey>(DetailsRepository.GetPrimaryKey(entity), EntityMessageType.Changed));
                }
                ReloadDetails();
            } catch (DbException e) {
                MessageBoxService.ShowMessage(e.ErrorMessage, e.ErrorCaption, MessageButton.OK, MessageIcon.Error);
            }
        }

        void OnMessage(EntityMessage<TEntity, TPrimaryKey> message) {
            if(message.MessageType == EntityMessageType.Changed && Entity != null && object.Equals(PrimaryKey, message.PrimaryKey)) {
                ReloadDetails();
            }
        }

        void ReloadDetails() {
            UnitOfWork = UnitOfWorkFactory.CreateUnitOfWork();
            OnParameterChanged(id);
            this.RaisePropertyChanged(x => x.DetailEntities);
        }
    }
    public class SelectDetailEntitiesViewModel<TEntity> where TEntity : class {
        public SelectDetailEntitiesViewModel(TEntity[] availableCourses) {
            AvailableEntities = availableCourses;
            SelectedEntities = new List<TEntity>();
        }
        public TEntity[] AvailableEntities { get; private set; }
        public List<TEntity> SelectedEntities { get; private set; }
    }
}

File: VB/Scaffolding.ManyToMany/Common/ViewModel/AddRemoveDetailEntitiesViewModel.vb
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO
Imports Scaffolding.ManyToMany.Common.DataModel
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.ComponentModel.DataAnnotations
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows

Namespace Scaffolding.ManyToMany.Common.ViewModel
    Public Class AddRemoveDetailEntitiesViewModel(Of TEntity As Class, TPrimaryKey, TDetailEntity As Class, TDetailPrimaryKey, TUnitOfWork As IUnitOfWork)
        Inherits SingleObjectViewModelBase(Of TEntity, TPrimaryKey, TUnitOfWork)

        Public Shared Function Create(ByVal unitOfWorkFactory As IUnitOfWorkFactory(Of TUnitOfWork), ByVal getRepositoryFunc As Func(Of TUnitOfWork, IRepository(Of TEntity, TPrimaryKey)), ByVal getDetailsRepositoryFunc As Func(Of TUnitOfWork, IRepository(Of TDetailEntity, TDetailPrimaryKey)), ByVal getDetailsFunc As Func(Of TEntity, ICollection(Of TDetailEntity)), ByVal id As TPrimaryKey) As AddRemoveDetailEntitiesViewModel(Of TEntity, TPrimaryKey, TDetailEntity, TDetailPrimaryKey, TUnitOfWork)
            Return ViewModelSource.Create(Function() New AddRemoveDetailEntitiesViewModel(Of TEntity, TPrimaryKey, TDetailEntity, TDetailPrimaryKey, TUnitOfWork)(unitOfWorkFactory, getRepositoryFunc, getDetailsRepositoryFunc, getDetailsFunc, id))
        End Function

        Private ReadOnly id As TPrimaryKey
        Private ReadOnly getDetailsRepositoryFunc As Func(Of TUnitOfWork, IRepository(Of TDetailEntity, TDetailPrimaryKey))
        Private ReadOnly getDetailsFunc As Func(Of TEntity, ICollection(Of TDetailEntity))

        <Required> _
        Protected Overridable ReadOnly Property DialogService() As IDialogService
            Get
                Return Nothing
            End Get
        End Property

        Private ReadOnly Property DetailsRepository() As IRepository(Of TDetailEntity, TDetailPrimaryKey)
            Get
                Return getDetailsRepositoryFunc(UnitOfWork)
            End Get
        End Property

        Public ReadOnly Property DetailEntities() As ICollection(Of TDetailEntity)
            Get
                Return If(Entity IsNot Nothing, getDetailsFunc(Entity), Enumerable.Empty(Of TDetailEntity)().ToArray())
            End Get
        End Property
        Private privateSelectedEntities As ObservableCollection(Of TDetailEntity)
        Public Property SelectedEntities() As ObservableCollection(Of TDetailEntity)
            Get
                Return privateSelectedEntities
            End Get
            Private Set(ByVal value As ObservableCollection(Of TDetailEntity))
                privateSelectedEntities = value
            End Set
        End Property

        Protected Sub New(ByVal unitOfWorkFactory As IUnitOfWorkFactory(Of TUnitOfWork), ByVal getRepositoryFunc As Func(Of TUnitOfWork, IRepository(Of TEntity, TPrimaryKey)), ByVal getDetailsRepositoryFunc As Func(Of TUnitOfWork, IRepository(Of TDetailEntity, TDetailPrimaryKey)), ByVal getDetailsFunc As Func(Of TEntity, ICollection(Of TDetailEntity)), ByVal id As TPrimaryKey)
            MyBase.New(unitOfWorkFactory, getRepositoryFunc, Nothing)
            Me.id = id
            Me.getDetailsRepositoryFunc = getDetailsRepositoryFunc
            Me.getDetailsFunc = getDetailsFunc
            SelectedEntities = New ObservableCollection(Of TDetailEntity)()
            ReloadDetails()
        End Sub

        Protected Overrides Sub OnInitializeInRuntime()
            MyBase.OnInitializeInRuntime()
            Messenger.Default.Register(Of EntityMessage(Of TEntity))(Me, AddressOf OnMessage)
        End Sub

        Public Sub AddDetailEntities()
            Dim availalbleEntities = DetailsRepository.GetEntities().ToList().Except(DetailEntities).ToArray()
            Dim selectEntitiesViewModel = New SelectDetailEntitiesViewModel(Of TDetailEntity)(availalbleEntities)
            If DialogService.ShowDialog(MessageBoxButton.OKCancel, "Select objects to add", selectEntitiesViewModel) = MessageBoxResult.OK AndAlso selectEntitiesViewModel.SelectedEntities.Any() Then

                For Each entity_Renamed In selectEntitiesViewModel.SelectedEntities
                    DetailEntities.Add(entity_Renamed)
                Next entity_Renamed
                SaveChangesAndNotify(selectEntitiesViewModel.SelectedEntities)
            End If
        End Sub

        Public Function CanAddDetailEntities() As Boolean
            Return Entity IsNot Nothing
        End Function

        Public Sub RemoveDetailEntities()
            If Not SelectedEntities.Any() Then
                Return
            End If

            For Each entity_Renamed In SelectedEntities
                DetailEntities.Remove(entity_Renamed)
            Next entity_Renamed
            SaveChangesAndNotify(SelectedEntities)
            SelectedEntities.Clear()
        End Sub

        Public Function CanRemoveDetailEntities() As Boolean
            Return Entity IsNot Nothing
        End Function

        Private Sub SaveChangesAndNotify(ByVal detailEntities As IEnumerable(Of TDetailEntity))
            Try
                UnitOfWork.SaveChanges()

                For Each entity_Renamed In detailEntities
                    Messenger.Default.Send(New EntityMessage(Of TDetailEntity)(entity_Renamed, EntityMessageType.Changed))
                Next entity_Renamed
                ReloadDetails()
            Catch e As DbException
                MessageBoxService.Show(e.ErrorMessage, e.ErrorCaption, MessageBoxButton.OK, MessageBoxImage.Error)
            End Try
        End Sub

        Private Sub OnMessage(ByVal message As EntityMessage(Of TEntity))
            If Entity IsNot Nothing AndAlso Object.Equals(GetPrimaryKey(Entity), GetPrimaryKey(message.Entity)) AndAlso message.MessageType = EntityMessageType.Changed Then
                ReloadDetails()
            End If
        End Sub

        Private Sub ReloadDetails()
            UnitOfWork = UnitOfWorkFactory.CreateUnitOfWork()
            OnParameterChanged(id)
            Me.RaisePropertyChanged(Function(x) x.DetailEntities)
        End Sub
    End Class
    Public Class SelectDetailEntitiesViewModel(Of TEntity As Class)
        Public Sub New(ByVal availableCourses() As TEntity)
            AvailableEntities = availableCourses
            SelectedEntities = New List(Of TEntity)()
        End Sub
        Private privateAvailableEntities As TEntity()
        Public Property AvailableEntities() As TEntity()
            Get
                Return privateAvailableEntities
            End Get
            Private Set(ByVal value As TEntity())
                privateAvailableEntities = value
            End Set
        End Property
        Private privateSelectedEntities As List(Of TEntity)
        Public Property SelectedEntities() As List(Of TEntity)
            Get
                Return privateSelectedEntities
            End Get
            Private Set(ByVal value As List(Of TEntity))
                privateSelectedEntities = value
            End Set
        End Property
    End Class
End Namespace
